docs/RFCS/20191209_optimistic_ddl.md

```diff
@@ -0,0 +1,650 @@
+
+#### Integer types
+
+![compatible-integer-types](../media/compatible-integer-types.png)
+
+TiDB does not support changing signedness.
+
+#### Sets and enums
+
+![compatible-sets-enums](../media/compatible-sets-enums.png)
+
+Changing enums and sets are compatible only if members are appended to the end.
+
+#### Other types
+
+The remaining (year, date, time, datetime, timestamp, float, double, decimal, bit, json) are independent and cannot be converted to or from other types.
+
+### Constraints and indices
+
+A table with less constraints are more compatible. Therefore, a DDL adding constraints should be executed last, while a DDL removing constraints should be executed first.
+
+#### Checks
+
+![compatible-check](../media/compatible-check.png)
+
+When encountering DDLs modifying checks, either propagate the DDL or drop it depending whether the check (with the same name) still exists in the joined schema.
+
+```sql
+alter table tbl01 add check (expr);
+alter table tbl01 drop check name;
+alter table tbl01 alter check name [not] enforced;
+```
+
+(Note that, in MySQL, “check” clauses in a “modify column” DDL is ignored.)
+
+#### Indices
+
+![compatible-indices](../media/compatible-indices.png)
+
+When encountering DDLs adding or removing indices, either propagate the DDL or drop it depending whether the index (with the same name) still exists in the joined schema. DM master should reply to DM worker as soon as possible, without waiting for TiDB to complete adding the indices.
+
+```sql
+alter table tbl01 add [spatial | fulltext | unique] key name (cols);
+alter table tbl01 alter check name [not] enforced;
+alter table tbl01 drop key name;
+create [spatial | fulltext | unique] index name on tbl01 (cols);
+drop index name on tbl01;
+alter table tbl01 modify column col02 col_spec unique;
+alter table tbl01 rename key old_name to new_name;
+```
+
+#### Foreign keys
+
+(TODO; Not supported by TiDB)
+
+### Adding and dropping columns
+
+Columns can be in one of three states:
+
+- Exists on all shards
+
+- Partially exists
+
+- Dropped from all shards
+
+![add-drop-column](../media/add-drop-column.png)
+
+Since DM master has a copy of all sharded schemas, it can simply count how many times this column exists on each shard to determine the column’s state in the joined schema.
+
+#### Adding column on first shard (dropped → partial)
+
+![compatible-add-column-on-first-shard](../media/compatible-add-column-on-first-shard.png)
+
+When adding a column which did not exist before, we should change the specification to include a default value if not given.
+
+| Type | Default |
+| ---- | ------- |
+| Any nullable type | null |
+| year | '0000' |
+| int | 0 |
+| date | '0000-00-00' |
+| float/double | 0.0 |
+| time | '00:00:00' |
+| decimal | 0 |
+| datetime | '0000-00-00 00:00:00' |
+| bit | 0 |
+| timestamp | '0000-00-00 00:00:00' |
+| char/varchar/text | '' |
+| enum/set | 0 |
+| binary/varbinary/blob | '' |
+| json | 'null' |
```

Comment:
in the original doc, the order is _column first_ but not _row first_.

Reply:
update in 2b44322